From a3de46fea214bbb059fc644c5b48e31d0bee45c6 Mon Sep 17 00:00:00 2001
From: Vladimir Oltean <vladimir.oltean@nxp.com>
Date: Tue, 28 Jan 2020 23:55:17 +0200
Subject: [PATCH 05/26] arm64: dts: fsl: ls1028a-rdb: Free up eno2/swp4 for
 802.1CB by enabling eno3/swp5

This patch moves the CPU port from swp4 to swp5 on the Felix switch, and
uses the 2.5 port pair which used to be for DSA tagging in plain
untagged mode.

The switch core for 802.1CB FRER uses the MAC table (FDB) to assign
packets to a Seamless Stream ID (SSID). In turn, the MAC table needs the
Analyzer module (ANA) to inspect the frames.

But traffic that passes to/from the DSA CPU port has the BYPASS flag set
in the injection header, which means "bypass the analyzer module".

So traffic originated from the CPU will not be correctly assigned to a
SSID for 802.1CB if it is coming from the port pair where DSA tagging is
enabled.

As a workaround, we enable the other port pair and use that for 802.1CB
traffic that needs to be terminated on the local CPU.

Before:                               After:

    eno2    eno3                         eno2      eno3
     |       |                   Untagged |         |
     | DSA   x disabled           traffic |         | DSA
     | tags  |                  including |         | tags
     |       |                    802.1CB |         |
    swp4    swp5                         swp4      swp5

Of course, in this context "untagged" means "not tagged with an
injection/extraction DSA tag", it has nothing to do with the VLAN or
802.1CB R-Tag.

Signed-off-by: Vladimir Oltean <vladimir.oltean@nxp.com>
---
 arch/arm64/boot/dts/freescale/fsl-ls1028a-rdb.dts | 13 +++++++++++++
 arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi    |  4 ++--
 2 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/fsl-ls1028a-rdb.dts b/arch/arm64/boot/dts/freescale/fsl-ls1028a-rdb.dts
index f147f83eccba..83226117046a 100644
--- a/arch/arm64/boot/dts/freescale/fsl-ls1028a-rdb.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-ls1028a-rdb.dts
@@ -214,6 +214,10 @@
 	status = "disabled";
 };
 
+&enetc_port3 {
+	status = "okay";
+};
+
 &enetc_mdio_pf3 {
 	qsgmii_phy1: ethernet-phy@4 {
 		reg = <0x10>;
@@ -257,6 +261,15 @@
 	managed = "in-band-status";
 };
 
+&switch_port4 {
+	/delete-property/ ethernet;
+};
+
+&switch_port5 {
+	status = "okay";
+	ethernet = <&enetc_port3>;
+};
+
 &sai4 {
 	status = "okay";
 };
diff --git a/arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi b/arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi
index d7939cb265c8..227be64f1c0b 100644
--- a/arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi
@@ -786,7 +786,7 @@
 						reg = <3>;
 					};
 					/* internal to-cpu ports */
-					port@4 {
+					switch_port4: port@4 {
 						reg = <4>;
 						ethernet = <&enetc_port2>;
 						phy-mode = "gmii";
@@ -796,7 +796,7 @@
 							full-duplex;
 						};
 					};
-					port@5 {
+					switch_port5: port@5 {
 						reg = <5>;
 						phy-mode = "gmii";
 						status = "disabled";
-- 
2.17.1

